pkg/prometheus/agent_http.go

```diff
@@ -0,0 +1,231 @@
+	r.HandleFunc("/agent/api/v1/configs/{name}", getConfig).Methods("GET")
+	r.HandleFunc("/agent/api/v1/config/{name}", putConfig).Methods("PUT", "POST")
+	r.HandleFunc("/agent/api/v1/config/{name}", deleteConfig).Methods("DELETE")
+}
+
+// WrapHandler is responsible for turning an APIHandler into an HTTP
+// handler by wrapping responses and writing them as JSON.
+func (a *Agent) WrapHandler(next APIHandler) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		resp, err := next(r)
+		if err != nil {
+			httpErr, ok := err.(*httpError)
+			var err error
+
+			if ok {
+				err = configapi.WriteError(w, httpErr.StatusCode, httpErr.Err)
+			} else {
+				err = configapi.WriteError(w, http.StatusInternalServerError, err)
+			}
+
+			if err != nil {
+				level.Error(a.logger).Log("msg", "failed writing error response to client", "err", err)
+			}
+			return
+		}
+
+		if err := configapi.WriteResponse(w, http.StatusOK, resp); err != nil {
+			level.Error(a.logger).Log("msg", "failed to write valid response", "err", err)
+		}
+	})
+}
+
+// ListConfigurations returns a list of the named configurations or all
+// configurations associated with the Prometheus agent.
+func (a *Agent) ListConfigurations(r *http.Request) (interface{}, error) {
+	vv, err := a.kv.List(r.Context(), "")
+	if err != nil {
+		level.Error(a.logger).Log("msg", "failed to list keys in kv store", "err", err)
+		return nil, err
+	}
+
+	return &configapi.ListConfigurationsResponse{Configs: vv}, nil
+}
+
+// GetConfiguration returns an existing named configuration.
+func (a *Agent) GetConfiguration(r *http.Request) (interface{}, error) {
+	ctype, err := configapi.ContentTypeFromRequest(r)
+	if err != nil {
+		return nil, err
+	}
+
+	configKey := getConfigName(r)
+	v, err := a.kv.Get(r.Context(), configKey)
+	if err != nil {
+		level.Error(a.logger).Log("msg", "error getting configuration from kv store", "err", err)
+		return nil, fmt.Errorf("error getting configuration: %w", err)
+	}
+
+	cfg, err := MarshalInstanceConfig(v.(*InstanceConfig), ctype)
+	if err != nil {
+		level.Error(a.logger).Log("msg", "error marshaling configuration", "err", err)
+		return nil, err
+	}
+
+	return &configapi.GetConfigurationResponse{Value: cfg}, nil
+}
+
+// PutConfiguration creates or updates a named configuration. Completely
+// overrides the previous configuration if it exists.
+func (a *Agent) PutConfiguration(r *http.Request) (interface{}, error) {
+	ctype, err := configapi.ContentTypeFromRequest(r)
+	if err != nil {
+		return nil, err
+	}
+
+	inst, err := UnmarshalInstanceConfig(r.Body, ctype)
+	if err != nil {
+		return nil, err
+	}
+	inst.Name = getConfigName(r)
+
+	var newConfig bool
+	err = a.kv.CAS(r.Context(), inst.Name, func(in interface{}) (out interface{}, retry bool, err error) {
+		// The configuration is new if there's no previous value from the CAS
+		newConfig = (in == nil)
+		return inst, false, nil
+	})
+
+	if err == nil {
+		if newConfig {
```

Comment:
I was expecting to see a `201` in this scenario. I noticed we always return 200 for all successful operations - should we expand on that?

Reply:
Hmm, I'm not sure how I would cleanly add in non-200 codes for successful operations here. Any ideas?

Reply:
One thing I might be able to do is have an `httpResponse` struct (kind of like the `httpError` one that wraps an `interface{}` but also lets you specify a custom response code. I think that'd make the most sense since 200 is going to be the most common reply here. WDYT?